### PR TITLE
Use `Write.UTxO` in `ErrBalanceTxInsufficientCollateralError`.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue

ADP-3184

## Description

This PR adjusts `ErrBalanceTxInsufficientCollateralError` to use `Write.UTxO` instead of the wallet primitive `UTxO` type.

### CODE REVIEW

File: lib/wallet-e2e/src/Cardano/Wallet/Spec/Effect/Trace.hs [hunk 1]
```diff
@@ -35,6 +35,7 @@ import Path.IO
 import Prelude hiding
     ( modify
     , runState
+    , trace
```
Comment (jonathanknowles) [inline lib/wallet-e2e/src/Cardano/Wallet/Spec/Effect/Trace.hs]: @Anviking 
We actually include a similar change in https://github.com/cardano-foundation/cardano-wallet/pull/4181 (https://github.com/cardano-foundation/cardano-wallet/pull/4181/commits/8e13bd4096d9d8c9ba1ddf69452f8e449df0f82c).

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 2]
```diff
@@ -3152,6 +3153,7 @@ constructSharedTransaction
     withWorkerCtx api wid liftE liftE $ \wrk -> do
         let db = wrk ^. dbLayer
             netLayer = wrk ^. networkLayer
+            trWorker :: Tracer IO W.WalletLog
```
Comment (Anviking) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:3161]: Are these required now?

Comment (jonathanknowles) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:3161]: > Are these required now?

Yes, it seems so.

I added type annotations/signatures one by one until I could get the whole codebase to compile.

File: lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs [hunk 3]
```diff
@@ -2122,6 +2122,10 @@ hasTotalCollateral (CardanoApi.Tx (CardanoApi.TxBody content) _) =
         CardanoApi.TxTotalCollateralNone -> False
         CardanoApi.TxTotalCollateral _ _ -> True
 
+-- | A type-constrained variant of 'it'.
+itIO :: String -> IO () -> Spec
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:2132]: Perhaps we could add a TODO here and link it to a ticket that reminds us to remove this when possible?

Comment (jonathanknowles) [reply lib/wallet/test/unit/Internal/Cardano/Write/Tx/BalanceSpec.hs:2132]: Fixed in 868dbcdb21085da7814b2c936a8d90cafa60e817.

File: lib/wallet/test/data/balanceTx/pingPong_2/golden [hunk 4]
```diff
@@ -39,26 +39,26 @@
...
+ 2.050000,ErrBalanceTxInsufficientCollateral (ErrBalanceTxInsufficientCollateralError {largestCombinationAvailable = UTxO (fromList [(TxIn (TxId {unTxId = SafeHash "3030303030303030303030303030303030303030303030303030303030303030"}) (TxIx 0),(Addr Testnet (KeyHashObj (KeyHash "b1e5e0fb74c86c801f646841e07cdb42df8b82ef3ce4e57cb5412e77")) StakeRefNull,MaryValue 2050000 (MultiAsset (fromList [])),NoDatum,SNothing))]), minimumCollateralAmount = Coin 3075000})
+ 2.100000,ErrBalanceTxInsufficientCollateral (ErrBalanceTxInsufficientCollateralError {largestCombinationAvailable = UTxO (fromList [(TxIn (TxId {unTxId = SafeHash "3030303030303030303030303030303030303030303030303030303030303030"}) (TxIx 0),(Addr Testnet (KeyHashObj (KeyHash "b1e5e0fb74c86c801f646841e07cdb42df8b82ef3ce4e57cb5412e77")) StakeRefNull,MaryValue 2100000 (MultiAsset (fromList [])),NoDatum,SNothing))]), minimumCollateralAmount = Coin 3150000})
```
Comment (Anviking) [inline lib/wallet/test/data/balanceTx/pingPong_2/golden:43]: 👍

**Review by Anviking — APPROVED** [review]
The fragile 8.10 type inference with `IsRecentEra` in its current state might cause confusion, but seems fine to me to go with for now and hope https://cardanofoundation.atlassian.net/browse/ADP-2353 or something similar will fix 👍

**Pull request merged**